src/hooks/use-form.js

```diff
@@ -0,0 +1,367 @@
+    case actionTypes.SUBMIT_FAILED:
+      return false;
+
+    default:
+      return state;
+  }
+}
+
+/**
+ * Form reducer.
+ */
+
+const formReducer = jsonSchema => (state: FormState, action: Action) => {
+  const values = valuesReducer(state.values, action);
+  const meta = metaReducer(state.meta, action);
+  const errors = errorsReducer(state.errors, action, jsonSchema, values);
+  const submitStatus = submitStatusReducer(state.submitStatus, action);
+  const isSubmitting = isSubmittingReducer(state.isSubmitting, action);
+
+  return {
+    errors,
+    isSubmitting,
+    meta,
+    submitStatus,
+    values
+  };
+};
+
+/**
+ * `Options` type.
+ */
+
+type Options = {
+  initialValues: Object,
+  jsonSchema: Object,
+  onSubmit: Submit,
+  onValuesChanged?: (formState: Object) => void
+};
+
+/**
+ * Export `useForm`.
+ */
+
+export default function useForm(options: Options) {
+  const { initialValues = {}, jsonSchema, onSubmit, onValuesChanged } = options;
+  const [state, dispatch] = useReducer(
+    formReducer(jsonSchema),
+    {
+      errors: {},
+      isSubmitting: false,
+      meta: {},
+      submitStatus: 'idle',
+      values: initialValues
+    }
+  );
+
+  const setFieldValue = useCallback((field: string, value: any) => {
+    dispatch({
+      payload: { field, value },
+      type: actionTypes.SET_FIELD_VALUE
+    });
+  }, []);
+
+  const blurField = useCallback((field: string) => {
+    dispatch({
+      payload: { field },
+      type: actionTypes.BLUR
+    });
+  }, []);
+
+  const focusField = useCallback((field: string) => {
+    dispatch({
+      payload: { field },
+      type: actionTypes.FOCUS
+    });
+  }, []);
+
+  const registerField = useCallback((field: string) => {
+    dispatch({
+      payload: { field },
+      type: actionTypes.REGISTER_FIELD
+    });
+  }, []);
+
+  const reset = useCallback(() => {
+    dispatch({
+      payload: { initialValues },
+      type: actionTypes.RESET
+    });
+  }, [initialValues]);
+
+  const submit = useCallback((event: ?SyntheticEvent<any>) => {
+    if (event && typeof event.preventDefault === 'function') {
+      event.preventDefault();
+    }
+
+    dispatch({
+      payload: {},
+      type: actionTypes.SUBMIT_START
+    });
+  }, []);
+
+  useEffect(() => {
+    if (onValuesChanged) {
+      onValuesChanged(state.values);
+    }
+  }, [state, onValuesChanged]);
+
+  useEffect(() => {
+    if (state.submitStatus !== 'submitStart') {
+      return;
+    }
+
+    dispatch({
+      payload: {},
+      type: actionTypes.SUBMITTING
+    });
+
+    if (Object.keys(state.errors).length > 0) {
+      dispatch({
+        payload: {},
+        type: actionTypes.SUBMIT_FAILED
+      });
+
+      return;
+    }
+
+    let canceled = false;
+
+    onSubmit(state.values, { reset }).then(result => {
+      if (!canceled) {
+        dispatch({
+          payload: {},
+          type: actionTypes.SUBMIT_END
+        });
+      }
+
+      return result;
+    });
```

Comment:
Shouldn't this have a `cath`?
Also, couldn't `finally` dispatch the `actionTypes.SUBMIT_END`?

Reply:
I added the rejection callback to the then to handle `onSubmit` errors.

Reply:
I also had to remove the cancellation logic, because the first dispatch in the effect was making the effect re-run and cancel the request.